c_binding/lsm_scsi.c

```diff
@@ -253,30 +205,31 @@ static int _sysfs_vpd83_naa_of_sd_name(char *err_msg, const char *sd_name,
         goto out;
     }
 
-    /*
-     * Check the existence of disk vis /sys/block/sdX folder.
-     */
-    snprintf(sysfs_blk_path, _MAX_SYSFS_BLK_PATH_STR_LEN,
-             _SYSFS_BLK_PATH_FORMAT, sd_name);
-    if (_file_is_exist(err_msg, sysfs_blk_path) == false) {
-        rc = LSM_ERR_NOT_FOUND_DISK;
+    udev = udev_new();
+    if (udev == NULL) {
+        rc = LSM_ERR_NO_MEMORY;
         goto out;
     }
 
-    snprintf(sysfs_path, _MAX_SYSFS_VPD83_PATH_STR_LEN,
-             _SYSFS_VPD83_PATH_FORMAT, sd_name);
-
-    rc = _sysfs_read_file(err_msg, sysfs_path, buff, &read_size,
-                          _MAX_VPD83_PAGE_LEN);
+    sd_udev = udev_device_new_from_subsystem_sysname(udev, "block", sd_name);
+    if (sd_udev == NULL) {
+        rc = LSM_ERR_NOT_FOUND_DISK;
+        goto out;
+    }
 
-    if (rc != LSM_ERR_OK)
+    value = udev_device_get_sysattr_value(sd_udev, "device/vpd_pg83");
```

Comment:
@cathay4t So, it turns out that libudev is entirely inconsistent in its vpd_pg83 attribute handling. Sometimes it returns a value:

```
>>> dev = pyudev.Device.from_path(context, '/sys/devices/pci0000:80/0000:80:02.0/0000:81:00.0/host2/target2:0:9/2:0:9:0')
>>> dev.attributes.get('vpd_pg83')
'\r\x83'
```

sometimes it returns an empty string:

```
>>> dev
Device(u'/sys/devices/pci0000:00/0000:00:03.0/0000:03:00.0/host1/target1:0:0/1:0:0:0/block/sda')
>>> dev.attributes.get('device/vpd_pg83')
''
```

Both of the corresponding files contain some data.

In contexts where there really is nothing to read it returns NULL.

From what you've told me, I guess that is all dependent on the value of the bytes that it encounters when reading the file. So, probably libudev has a pretty pervasive problem when reading binary data from attributes.

My understanding is that udevadm shouldn't be showing vpd_pg83 values at alll, because --attribute-walk is supposed to only show those textual attributes that can be matched against udev rules. It shows the value because it is incorrectly read and therefore interpreted as text.

I think I'll file a broader bug against libudev about this inconsistancy in reading non-textual sysfs attributes, and see what happens.

But, I guess that what we may know at this point is that libudev's reading of all non-text attributes is likely to be unreliable.